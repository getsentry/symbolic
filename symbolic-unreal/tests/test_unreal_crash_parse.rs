use std::fs::File;
use std::io::Read;

use symbolic_testutils::fixture;
use symbolic_unreal::{Unreal4Crash, Unreal4Error, Unreal4FileType};

#[cfg(test)]
use similar_asserts::assert_eq;

fn get_unreal_crash() -> Result<Unreal4Crash, Unreal4Error> {
    let mut file = File::open(fixture("unreal/unreal_crash")).expect("example file opens");
    let mut file_content = Vec::new();
    file.read_to_end(&mut file_content).expect("fixture file");
    Unreal4Crash::parse(&file_content)
}

fn get_unreal_apple_crash() -> Result<Unreal4Crash, Unreal4Error> {
    let mut file = File::open(fixture("unreal/unreal_crash_apple")).expect("example file opens");
    let mut file_content = Vec::new();
    file.read_to_end(&mut file_content).expect("fixture file");
    Unreal4Crash::parse(&file_content)
}

#[test]
fn test_load_unreal_crash() {
    get_unreal_crash().expect("crash file loaded");
}

#[test]
fn test_get_minidump_slice() {
    let ue4_crash = get_unreal_crash().expect("test crash file loads");

    let minidump_file = ue4_crash
        .file_by_type(Unreal4FileType::Minidump)
        .expect("expected minidump file exists");

    assert_eq!(minidump_file.data().len(), 410_700);

    let crash_file = ue4_crash
        .native_crash()
        .expect("expected native crash exists");

    assert_eq!(minidump_file.index(), crash_file.index());

    assert!(ue4_crash
        .file_by_type(Unreal4FileType::AppleCrashReport)
        .is_none());
}

#[test]
fn test_get_apple_crash_report() {
    let ue4_crash = get_unreal_apple_crash().expect("test crash file loads");

    let apple_file = ue4_crash
        .file_by_type(Unreal4FileType::AppleCrashReport)
        .expect("expected apple crash report file exists");

    assert_eq!(apple_file.data().len(), 91_392);

    let crash_file = ue4_crash
        .native_crash()
        .expect("expected native crash exists");

    assert_eq!(apple_file.index(), crash_file.index());

    assert!(ue4_crash.file_by_type(Unreal4FileType::Minidump).is_none());
}

#[test]
#[cfg(feature = "serde")]
fn test_contexts_runtime_properties() {
    let ue4_crash = get_unreal_crash().expect("test crash file loads");

    let ue4_context = ue4_crash
        .context()
        .expect("no errors parsing the context file")
        .expect("context file exists in sample crash");

    let runtime_properties = ue4_context
        .runtime_properties
        .expect("runtime properties exist within sample crash");

    assert_eq!(
        "UE4CC-Windows-379993BB42BD8FBED67986857D8844B5_0000",
        runtime_properties.crash_guid.as_ref().expect("crash guid")
    );

    insta::assert_yaml_snapshot!(&runtime_properties);
}

#[test]
fn test_contexts_platform_properties() {
    let ue4_crash = get_unreal_crash().expect("test crash file loads");

    let ue4_context = ue4_crash
        .context()
        .expect("no errors parsing the context file")
        .expect("context file exists in sample crash");

    let platform_properties = ue4_context
        .platform_properties
        .expect("platform properties exist within sample crash");

    assert_eq!(
        platform_properties
            .is_windows
            .expect("sample contains value as 1 for true"),
        true
    );

    assert_eq!(
        platform_properties
            .callback_result
            .expect("sample contains value 0"),
        0
    );
}

#[test]
fn test_files_api() {
    let ue4_crash = get_unreal_crash().expect("test crash file loads");

    assert_eq!(ue4_crash.file_count(), 4);
    assert_eq!(ue4_crash.files().size_hint(), (4, Some(4)));

    assert_eq!(
        ue4_crash.file_by_index(0).expect("File exists").name(),
        "CrashContext.runtime-xml"
    );
    assert_eq!(
        ue4_crash.file_by_index(1).expect("File exists").name(),
        "CrashReportClient.ini"
    );
    assert_eq!(
        ue4_crash.file_by_index(2).expect("File exists").name(),
        "MyProject.log"
    );
    assert_eq!(
        ue4_crash.file_by_index(3).expect("File exists").name(),
        "UE4Minidump.dmp"
    );

    let context_file = ue4_crash.file_by_index(0).expect("xml file in pos 0");
    let xml = context_file.data();

    assert_eq!(xml[0] as char, '<');
    // there are two line breaks after closing tag:
    assert_eq!(xml[xml.len() - 3] as char, '>');
}

#[test]
fn test_get_logs() {
    let ue4_crash = get_unreal_crash().expect("test crash file loads");
    let limit = 100;
    let logs = ue4_crash.logs(limit).expect("log file");

    assert_eq!(logs.len(), limit);
    assert_eq!(
        logs[1].timestamp.expect("timestamp").to_rfc3339(),
        "2018-10-29T16:56:37+00:00"
    );
    assert_eq!(
        logs[0].component.as_ref().expect("component"),
        "LogD3D11RHI"
    );
    assert_eq!(logs[0].message, "Chosen D3D11 Adapter: 0");

    assert_eq!(
        logs[99].timestamp.expect("timestamp").to_rfc3339(),
        "2018-10-29T16:56:38+00:00"
    );
    assert_eq!(
        logs[99].component.as_ref().expect("component"),
        "LogWindows"
    );
    assert_eq!(
        logs[99].message,
        "Windows GetLastError: The operation completed successfully. (0)"
    );
}
