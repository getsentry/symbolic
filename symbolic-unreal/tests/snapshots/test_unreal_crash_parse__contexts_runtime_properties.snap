---
source: symbolic-unreal/tests/test_unreal_crash_parse.rs
expression: "&runtime_properties"
---
crash_guid: UE4CC-Windows-379993BB42BD8FBED67986857D8844B5_0000
process_id: 20944
is_internal_build: false
is_source_distribution: false
is_assert: true
is_ensure: false
crash_type: Assert
seconds_since_start: 0
game_name: UE4-MyProject
executable_name: MyProject
build_configuration: Development
platform_name: WindowsNoEditor
engine_mode: Game
engine_version: 4.20.3-4369336+++UE4+Release-4.20
language_lcid: 1033
app_default_locate: en-US
build_version: ++UE4+Release-4.20-CL-4369336
is_ue4_release: false
username: bruno
base_dir: "C:/Users/bruno/OneDrive/Documents/build-windows-64/WindowsNoEditor/MyProject/Binaries/Win64/"
root_dir: "C:/Users/bruno/OneDrive/Documents/build-windows-64/WindowsNoEditor/"
machine_id: 6B5EF42F48669C4D7DFE11B7F59BB6AC
login_id: 6b5ef42f48669c4d7dfe11b7f59bb6ac
epic_account_id: 2e7d369327054a448be6c8d3601213cb
legacy_call_stack: "MyProject!FDebug::AssertFailed()\nMyProject!AActor::BeginPlay()\nMyProject!AActor::DispatchBeginPlay()\nMyProject!AWorldSettings::NotifyBeginPlay()\nMyProject!AGameStateBase::HandleBeginPlay()\nMyProject!UWorld::BeginPlay()\nMyProject!UEngine::LoadMap()\nMyProject!UEngine::Browse()\nMyProject!UGameInstance::StartGameInstance()\nMyProject!FEngineLoop::Init()\nMyProject!GuardedMain()\nMyProject!GuardedMainWrapper()\nMyProject!WinMain()\nMyProject!__scrt_common_main_seh() [f:\\dd\\vctools\\crt\\vcstartup\\src\\startup\\exe_common.inl:288]\nkernel32\nntdll"
portable_call_stack: "0x00000000ffffffff + ffffffff 0x00000000ffffffff + ffffffff"
error_message: "Assertion failed: false [File:C:\\Users\\bruno\\OneDrive\\Documents\\Unreal Projects\\MyProject\\Source\\MyProject\\NewActorComponent.cpp] [Line: 21] test"
misc_number_of_cores: 6
misc_number_of_cores_inc_hyperthread: 6
misc_cpu_vendor: GenuineIntel
misc_cpu_brand: Intel(R) Core(TM) i5-8600K CPU @ 3.60GHz
misc_primary_gpu_brand: NVIDIA GeForce GTX 1080
misc_os_version_major: Windows 10
memory_stats_total_physical: 17112477696
memory_stats_total_virtual: 140737488224256
memory_stats_page_size: 4096
memory_stats_total_phsysical_gb: 16
time_of_crash: 636764289988050000
crash_reporter_client_version: "1.0"
modules: "C:\\Users\\bruno\\OneDrive\\Documents\\build-windows-64\\WindowsNoEditor\\MyProject\\Binaries\\Win64\\MyProject.exe\nC:\\Users\\bruno\\OneDrive\\Documents\\build-windows-64\\WindowsNoEditor\\Engine\\Binaries\\ThirdParty\\Vorbis\\Win64\\VS2015\\libvorbis_64.dll\nC:\\Users\\bruno\\OneDrive\\Documents\\build-windows-64\\WindowsNoEditor\\Engine\\Binaries\\ThirdParty\\PhysX3\\Win64\\VS2015\\APEX_LegacyPROFILE_x64.dll\nC:\\Users\\bruno\\OneDrive\\Documents\\build-windows-64\\WindowsNoEditor\\Engine\\Binaries\\ThirdParty\\PhysX3\\Win64\\VS2015\\PhysX3PROFILE_x64.dll\nC:\\Users\\bruno\\OneDrive\\Documents\\build-windows-64\\WindowsNoEditor\\Engine\\Binaries\\ThirdParty\\PhysX3\\Win64\\VS2015\\APEX_ClothingPROFILE_x64.dll\nC:\\Users\\bruno\\OneDrive\\Documents\\build-windows-64\\WindowsNoEditor\\Engine\\Binaries\\ThirdParty\\Oculus\\OVRPlugin\\OVRPlugin\\Win64\\OVRPlugin.dll\nC:\\Users\\bruno\\OneDrive\\Documents\\build-windows-64\\WindowsNoEditor\\Engine\\Binaries\\ThirdParty\\PhysX3\\Win64\\VS2015\\APEXFrameworkPROFILE_x64.dll\nC:\\Users\\bruno\\OneDrive\\Documents\\build-windows-64\\WindowsNoEditor\\Engine\\Binaries\\ThirdParty\\PhysX3\\Win64\\VS2015\\PhysX3CommonPROFILE_x64.dll\nC:\\Users\\bruno\\OneDrive\\Documents\\build-windows-64\\WindowsNoEditor\\Engine\\Binaries\\ThirdParty\\PhysX3\\Win64\\VS2015\\NvClothPROFILE_x64.dll\nC:\\Users\\bruno\\OneDrive\\Documents\\build-windows-64\\WindowsNoEditor\\Engine\\Binaries\\ThirdParty\\PhysX3\\Win64\\VS2015\\PhysX3CookingPROFILE_x64.dll\nC:\\Users\\bruno\\OneDrive\\Documents\\build-windows-64\\WindowsNoEditor\\Engine\\Binaries\\ThirdParty\\PhysX3\\Win64\\VS2015\\PxPvdSDKPROFILE_x64.dll\nC:\\Users\\bruno\\OneDrive\\Documents\\build-windows-64\\WindowsNoEditor\\Engine\\Binaries\\ThirdParty\\Ogg\\Win64\\VS2015\\libogg_64.dll\nC:\\Users\\bruno\\OneDrive\\Documents\\build-windows-64\\WindowsNoEditor\\Engine\\Binaries\\ThirdParty\\PhysX3\\Win64\\VS2015\\PxFoundationPROFILE_x64.dll\nC:\\Users\\bruno\\OneDrive\\Documents\\build-windows-64\\WindowsNoEditor\\Engine\\Binaries\\ThirdParty\\OpenVR\\OpenVRv1_0_11\\Win64\\openvr_api.dll\nC:\\Users\\bruno\\OneDrive\\Documents\\build-windows-64\\WindowsNoEditor\\Engine\\Binaries\\ThirdParty\\NVIDIA\\NVaftermath\\Win64\\GFSDK_Aftermath_Lib.x64.dll\nC:\\Users\\bruno\\OneDrive\\Documents\\build-windows-64\\WindowsNoEditor\\Engine\\Binaries\\ThirdParty\\Vorbis\\Win64\\VS2015\\libvorbisfile_64.dll"
custom:
  CommandLine: MyProject
  CrashDumpMode: "0"
  CrashVersion: "3"
  DeploymentName: ""
  EngineModeEx: Unset
  GameSessionID: ""
  IsPerforceBuild: "false"
  MemoryStats.AvailablePhysical: "10417119232"
  MemoryStats.AvailableVirtual: "140732265930752"
  MemoryStats.OOMAllocationAlignment: "0"
  MemoryStats.OOMAllocationSize: "0"
  MemoryStats.PeakUsedPhysical: "285224960"
  MemoryStats.PeakUsedVirtual: "402124800"
  MemoryStats.UsedPhysical: "244899840"
  MemoryStats.UsedVirtual: "392978432"
  MemoryStats.bIsOOM: "0"
  PlatformFullName: "Win64 [Windows 10  64b]"
  PlatformNameIni: Windows
  SourceContext: ""
  UserActivityHint: ""
